src/NHibernate/Cache/CacheFactory.cs

```diff
@@ -30,27 +30,50 @@ public static class CacheFactory
 		/// <param name="settings">Used to retrieve the global cache region prefix.</param>
 		/// <param name="properties">Properties the cache provider can use to configure the cache.</param>
 		/// <returns>An <see cref="ICacheConcurrencyStrategy"/> to use for this object in the <see cref="ICache"/>.</returns>
-		public static ICacheConcurrencyStrategy CreateCache(string usage, string name, bool mutable, Settings settings,
-		                                                    IDictionary<string,string> properties)
+		// Since v5.1
+		[Obsolete("Please use overload with an ICache parameter.")]
+		public static ICacheConcurrencyStrategy CreateCache(
+			string usage,
+			string name,
+			bool mutable,
+			Settings settings,
+			IDictionary<string, string> properties)
 		{
-			if (usage == null || !settings.IsSecondLevelCacheEnabled) return null; //no cache
+			var cache = CreateCache(
+				usage, name, settings,
+				(r, u) => settings.CacheProvider.BuildCache(r, properties));
 
-			string prefix = settings.CacheRegionPrefix;
-			if (prefix != null) name = prefix + '.' + name;
+			if (cache != null && mutable && usage == ReadOnly)
+				Log.Warn("read-only cache configured for mutable: {0}", name);
+
+			return cache;
+		}
+
+		/// <summary>
+		/// Creates an <see cref="ICacheConcurrencyStrategy"/> from the parameters.
+		/// </summary>
+		/// <param name="usage">The name of the strategy that <see cref="ICacheProvider"/> should use for the class.</param>
+		/// <param name="name">The name of the cache region the strategy is being created for.</param>
+		/// <param name="settings">Used to retrieve the global cache region prefix.</param>
+		/// <param name="regionAndUsageCacheGetter">The delegate for obtaining the <see cref="ICache" /> to use for the region.</param>
+		/// <returns>An <see cref="ICacheConcurrencyStrategy"/> to use for this object in the <see cref="ICache"/>.</returns>
+		public static ICacheConcurrencyStrategy CreateCache(
+			string usage,
+			string name,
+			Settings settings,
+			Func<string, string, ICache> regionAndUsageCacheGetter)
+		{
+			if (usage == null || !settings.IsSecondLevelCacheEnabled) return null; //no cache
 
-			if (log.IsDebugEnabled())
+			if (Log.IsDebugEnabled())
 			{
-				log.Debug("cache for: {0} usage strategy: {1}", name, usage);
+				Log.Debug("cache for: {0} usage strategy: {1}", name, usage);
 			}
 
 			ICacheConcurrencyStrategy ccs;
 			switch (usage)
 			{
 				case ReadOnly:
-					if (mutable)
-					{
-						log.Warn("read-only cache configured for mutable: {0}", name);
-					}
```

Comment:
Moved into session factory. Only the session factory can reliably issue this warning, since a cache concurrency may be built for an entity then reused for another one having a different mutability.